### ISSUE TITLE
Tracking: Third party NPM replacements

### ISSUE THREAD

**Issue by hansl** [angular/angular-cli#3886]
In the recent weeks, multiple have requested support for the following two NPM replacements:

- [ ] `yarn`, which has a lot of speed improvements. Many users reported problems using yarn with the CLI in some occasions.
- [ ] `cnpm`, which uses an NPM cache specially made for China. It's not useful for anyone outside of China itself.

What I suggest is embracing the fact that users can have an `angular-cli.json` file in their home folder to specify which NPM replacement they want to use.

Something like adding to the configuration:

```json
{ // ...
  "packageManager": "npm"  // or "cnpm" or "yarn"
}
```

And having special code for some of them when the user creates a new package.

I think this is the cleanest way to do it. If we document it well, users won't have a problem. They can even replace its value per-project.

cc @IgorMinar

**Comment by Meligy** [comment]
Maybe for now just add a note about the `ski-npm` flag advertising it as "for users who want to use alternate package managers like `yarn` or `cnpm`".

**Comment by hansl** [comment]
Discussing this more with @IgorMinar, we came to the point where the following would be acceptable:

- On `ng new`, if we detect either `yarn` or `cnpm`, we output a warning message to users, telling them they can `ng set --global packageManager=yarn` to use it globally with new projects.
- Users can then set it globally and recreate the project, or `yarn install` locally instead.

NPM is only used when running `init` and `new`, and since `init` is used by `new` we should just put the warning on `init`.

**Comment by Meligy** [comment]
For those subscribing to this issue, particularly the yarn crowd, here's a 6 minutes video that shows how yarn works with Angular CLI today:

https://youtu.be/jqvk1qBb4CY

**Comment by christopherthielen** [comment]
tl;dw:

```
ng new test-project -sn
cd test-project
yarn
git add yarn.lock
```

**Comment by Meligy** [comment]
For Yarn in particular, you also need to ensure you have 0.19 or higher. We had people report issues in the past where upgrading from 0.17 to 0.19 fixed it for them, and I'm aware of issues with global packages in general that were show stoppers to me and others, and were fixed in 0.19.

**Comment by Meligy** [comment]
I've added a written version with explanation here:

https://www.gurustop.net/blog/2017/01/26/use-yarn-npm-with-angular-cli

**Comment by sivabudh** [comment]
TL;DR: 

```
ng set --global packageManager=yarn
ng new <your-app-name>
```

**Comment by RamVellanki** [comment]
Hi, I tried using yarn by doing `ng set --global packagemanager=yarn` but I am getting the following error:
```
Successfully initialized git.
Installing packages for tooling via yarn.
warning You don't appear to have an internet connection. Try the --offline flag to use the cache for registry queries.
error An unexpected error occurred: "https://registry.yarnpkg.com/@angular%2fcommon: getaddrinfo ENOENT registry.yarnpkg.com:443".
```
